### PR TITLE
Remove some unnecessary tf.casts

### PR DESCRIPTION
**Pull request by uri-granta**
**Related issue(s)/PRs:** 

## Summary

Remove some unnecessary casting, in particular in the various samplers.

**Fully backwards compatible:** yes

## PR checklist

- [ ] The quality checks are all passing
- [ ] The bug case / new feature is covered by tests
- [ ] Any new features are well-documented (in docstrings or notebooks)

### CODE REVIEW

File: trieste/acquisition/function/active_learning.py [hunk 1]
```diff
@@ -222,7 +222,7 @@ def acquisition(x: TensorType) -> TensorType:
         t = (threshold - mean) / stdev
         t_plus = t + alpha
         t_minus = t - alpha
-        normal = tfp.distributions.Normal(tf.cast(0, x.dtype), tf.cast(1, x.dtype))
+        normal = tfp.distributions.Normal(tf.constant(0, x.dtype), tf.constant(1, x.dtype))
```
Comment (hstojic) [inline trieste/acquisition/function/active_learning.py:225]: why is using constant better than casting?

Comment (uri-granta) [reply trieste/acquisition/function/active_learning.py:225]: tf.cast should be used for Tensors, tf.constant for static, non-Tensor types, and tf.convert_to_tensor for any mix of Tensor-like objects. Also, it reads better here, and makes testing for unnecessary casts easier.

File: tests/unit/models/gpflux/test_sampler.py [hunk 2]
```diff
@@ -151,7 +152,17 @@ def test_dgp_reparam_sampler_sample_is_repeatable(
 
     sampler = DeepGaussianProcessReparamSampler(100, model)
     xs = tf.random.uniform([100, 2], minval=-10.0, maxval=10.0, dtype=tf.float64)[:, None, :]
-    npt.assert_allclose(sampler.sample(xs), sampler.sample(xs))
+
+    # also check there are no unnecessary casts from float64 to float32 or vice versa
```
Comment (uri-granta) [inline tests/unit/models/gpflux/test_sampler.py:156]: (this should be a separate test but the sampler tests are currently very slow due to training; I'll try to get them to use a session-scoped fixture in a separate PR)

File: trieste/acquisition/function/active_learning.py (hunk 1 above)
Comment (hstojic) [reply trieste/acquisition/function/active_learning.py:225]: thanks for the education :)

File: tests/unit/models/gpflux/test_sampler.py (hunk 2 above)
Comment (hstojic) [reply tests/unit/models/gpflux/test_sampler.py:156]: yes, that sounds good

File: tests/unit/models/gpflow/test_sampler.py [hunk 3]
```diff
@@ -350,6 +352,29 @@ def test_batch_reparametrization_sampler_samples_are_repeatable(qmc: bool, qmc_s
...
+    def patched_tf_cast(x: TensorType, dtype: tf.DType) -> TensorType:
+        # ensure there are no unnecessary casts from float64 to float32 or vice versa
+        if isinstance(x, tf.Tensor) and x.dtype in (tf.float32, tf.float64) and x.dtype != dtype:
+            raise ValueError(f"unexpected cast: {x} to {dtype}")
+        return original_tf_cast(x, dtype)
+
+    with patch("tensorflow.cast", side_effect=patched_tf_cast):
+        samples = sampler.sample(xs)
+        assert samples.dtype is dtype
+        npt.assert_allclose(samples, sampler.sample(xs))
```
Comment (hstojic) [inline tests/unit/models/gpflow/test_sampler.py:375]: neat test

File: tests/unit/models/keras/test_sampler.py [hunk 4]
```diff
@@ -74,7 +75,16 @@ def test_ensemble_trajectory_sampler_returns_trajectory_function_with_correctly_
...
+    original_tf_cast = tf.cast
+
+    def patched_tf_cast(x: TensorType, dtype: tf.DType) -> TensorType:
+        # ensure there are no unnecessary casts from float64 to float32 or vice versa
+        if isinstance(x, tf.Tensor) and x.dtype in (tf.float32, tf.float64) and x.dtype != dtype:
+            raise ValueError(f"unexpected cast: {x} to {dtype}")
+        return original_tf_cast(x, dtype)
+
+    with patch("tensorflow.cast", side_effect=patched_tf_cast):
+        assert trajectory(test_data).shape == (num_evals, batch_size, num_outputs)
```
Comment (hstojic) [inline tests/unit/models/keras/test_sampler.py]: why not testing the dtype of the output?
for DeepGPs perhaps not necessary as we should always use 64 there, but here we could conceivably use both...

**Review by hstojic — COMMENTED** [review]
some additional comments, see below...

we should probably also test that output of certain methods in models is consistent, i.e. dtype is as expected - but this can be done in a separate PR 

an additional question, acq fnc optimizer is not doing any conversion/hardcoding?

File: tests/unit/models/keras/test_sampler.py (hunk 4 above)
Comment (uri-granta) [reply tests/unit/models/keras/test_sampler.py]: So it turns out that the hidden layers in DeepEnsemble meant that we were always using float32 for the predictions. I've tentatively changed it so it uses the dtypes in the KerasEnsemble networks. Does it look ok?

**Review by hstojic — APPROVED** [review]
looks good

**Pull request merged**